keps/sig-auth/2579-psp-replacement/README.md

```diff
@@ -0,0 +1,1001 @@
+
+~~Using `deny` for the enforcing level would be more consistent with audit/warn (deny above), but is
+a counter-intuitive user experience. One would expect `deny = privileged` to deny privileged pods,
+but it really means allow privileged pods (hence using `allow` for the key).~~ (Rejected)
+
+Another option is to change the `audit` and `warn` keys to something meaning "don't audit/warn at this
+level and bellow", but we haven't thought of a good name. The best alternative we've come up with is
+`audit-above` and `warn-above`, but that is cumbersome.
+
+A final option is to change `allow` to `enforce`, which is more ambiguous. The ambiguity makes the
+inconsistency with audit & warn less obvious, but doesn't help the user experience.
+
+<<[/UNRESOLVED]>>
+
+There are several reasons for controlling the policy directly through namespace labels, rather than
+through a separate object:
+
+- Using labels enables various workflows around policy management through kubectl, for example
+  issuing queries like `kubectl get namespaces -l
+  <prefix>/allow-version!=v1.22` to find namespaces where the enforcing
+  policy isn't pinned to the most recent version.
+- Keeping the options on namespaces allows atomic create-and-set-policy, as opposed to creating a
+  namespace and then creating a second object inside the namespace.
+- Policies settings are per-namespace singletons, and singleton objects are not well supported in
+  Kubernetes.
+- Labels are lighter-weight than namespace fields, making it clear that this is a policy layer on
+  top of namespaces, not inherent to namespaces itself.
+
+### Validation
+
+The following restrictions are placed (by the admission plugin) on the policy namespace labels:
+
+1. Unknown labels with the pod security prefix (TBD) are rejected, e.g. `<prefix>/foo-bar`
+2. Policy level must be one of: `privileged`, `baseline`, `restricted`
+3. Version values must be match `(latest|v[0-9]+\.[0-9]+`. That is, one of:
+    1. `latest`
+    2. `vMAJOR.MINOR` (e.g. `v1.21`)
+
+Enforcement is best effort, and invalid labels that pre-existed the admission controller enablement
+are ignored. Updates to a previously invalid label are only allowed if the new value is valid.
+
+### Versioning
+
+A specific version can be supplied for each enforcement mode. The version pins the policy to the
+version that was defined at that kubernetes version. The default version is `latest`, which can be
+provided to explicitly use the latest definition of the policy. There is some nuance to how versions
+other than latest are applied:
+
+- If the constrained pod field has not changed since the pinned version, the policy is applied as
+  originally specified.
+- The privileged profile always means fully unconstrained and is effectively unversioned (specifying
+  a version is allowed but ignored).
+- Specifying a version more recent than the current Kubernetes version is allowed (for rollback &
+  version skew reasons), but is treated as `latest`.
+
```

Comment:
I'm assuming that “more recent” is as per is SemVer rather than calendar date of release.

Reply:
yes, semver

Reply:
it's not technically semver, since we're not including a patch version (just MAJOR.MINOR). In practice that tracks with the calendar, at least until we release 2.0 (at which point we might need to put more thought into this).